src/interfaces/python/swig_typemaps.i

```diff
@@ -824,71 +633,129 @@ static bool spvector_to_numpy(PyObject* &obj, SGSparseVector<type> sg_vector, in
-#endif
-TYPEMAP_IN_SGVECTOR(uint8_t,       NPY_UINT8)
-TYPEMAP_IN_SGVECTOR(int16_t,       NPY_INT16)
-TYPEMAP_IN_SGVECTOR(uint16_t,      NPY_UINT16)
-TYPEMAP_IN_SGVECTOR(int32_t,       NPY_INT32)
-TYPEMAP_IN_SGVECTOR(uint32_t,      NPY_UINT32)
-TYPEMAP_IN_SGVECTOR(int64_t,       NPY_INT64)
-TYPEMAP_IN_SGVECTOR(uint64_t,      NPY_UINT64)
-TYPEMAP_IN_SGVECTOR(float32_t,     NPY_FLOAT32)
-TYPEMAP_IN_SGVECTOR(float64_t,     NPY_FLOAT64)
-TYPEMAP_IN_SGVECTOR(floatmax_t,    NPY_LONGDOUBLE)
-TYPEMAP_IN_SGVECTOR(complex128_t,   NPY_CDOUBLE)
-TYPEMAP_IN_SGVECTOR(PyObject,      NPY_OBJECT)
-
-#undef TYPEMAP_IN_SGVECTOR
-
-/* One dimensional output arrays */
-%define TYPEMAP_OUT_SGVECTOR(type,typecode)
-%typemap(out) shogun::SGVector<type>
+// tell swig about the traits of the type to be used in stl containers
+%fragment(SWIG_Traits_frag(shogun::SGVector<type>), "header",
+          fragment="StdTraits",
+          fragment=SWIG_From_frag(shogun::SGVector<type>),
+          fragment=SWIG_AsVal_frag(shogun::SGVector<type>),
+          fragment=SWIG_From_frag(std::string),
+          fragment=SWIG_AsVal_frag(std::string))
 {
-    if (!vector_to_numpy($result, $1, typecode))
-        SWIG_fail;
+namespace swig
+{
+    template <>
+    struct traits<shogun::SGVector<type>>
+    {
+        typedef value_category category;
+        static const char* type_name()
+        {
+            return "shogun::SGVector<" #type ">";
+        }
+    };
+
+    template <>
+    struct traits_asval<shogun::SGVector<type>>
+    {
+        typedef shogun::SGVector<type> value_type;
+
+        static int asval(SWIG_Object obj, value_type *val)
+        {
+            if constexpr (!std::is_same<type, char>::value)
+            {
+                return SWIG_AsVal_dec(shogun::SGVector<type>)(obj, val);
+            }
+            else
+            {
+                // treat SGVector<char> as a string
+                std::string str;
+                int result = SWIG_AsVal_dec(std::string)(obj, &str);
+                if (SWIG_IsOK(result) && val)
+                {
+                    const char* cstr = str.c_str();
+                    *val = shogun::SGVector<type>(cstr, cstr + str.length());
+                }
+                return result;
+            }
+        }
+    };
+
+    template <>
+    struct traits_from<shogun::SGVector<type>>
+    {
+        typedef shogun::SGVector<type> value_type;
+
+        static SWIG_Object from(const value_type& val)
+        {
+            if constexpr (!std::is_same<type, char>::value)
+            {
+                return SWIG_From(shogun::SGVector<type>)(val);
+            }
+            else
+            {
+                // treat SGVector<char> as a string
+                std::stringstream ss;
+                ss.write((char *)val.vector, val.vlen);
+                return SWIG_From_dec(std::string)(ss.str());
+            }
+        }
+    };
+}
 }
+
+%typemaps_asvalfromn(SWIG_TYPECHECK_POINTER, shogun::SGVector<type>);
```

Comment:
this macro defines the "in" and "out" typemaps using the fragment conventions I used above